plugin/trino-loki/src/main/java/io/trino/plugin/loki/LokiTableFunction.java

```diff
@@ -0,0 +1,162 @@
+
+    private LokiMetadata metadata;
+
+    public LokiTableFunction(LokiMetadata metadata)
+    {
+        super(
+                SCHEMA_NAME,
+                NAME,
+                List.of(
+                        ScalarArgumentSpecification.builder()
+                                .name("QUERY")
+                                .type(VarcharType.VARCHAR)
+                                .build(),
+                        ScalarArgumentSpecification.builder()
+                                .name("START")
+                                .type(VarcharType.VARCHAR) // TODO: figure our if this could be a timestamp.
+                                .defaultValue(Slices.utf8Slice(""))
+                                .build(),
+                        ScalarArgumentSpecification.builder()
+                                .name("END")
+                                .type(VarcharType.VARCHAR)
+                                .defaultValue(Slices.utf8Slice(""))
+                                .build()),
+                GENERIC_TABLE);
+
+        this.metadata = metadata;
+    }
+
+    @Override
+    public TableFunctionAnalysis analyze(ConnectorSession session, ConnectorTransactionHandle transaction, Map<String, Argument> arguments, ConnectorAccessControl accessControl)
+    {
+        var argument = (ScalarArgument) arguments.get("QUERY");
+        String query = ((Slice) argument.getValue()).toStringUtf8();
+
+        String startArgument = ((Slice) ((ScalarArgument) arguments.get("START")).getValue()).toStringUtf8();
+        String endArgument = ((Slice) ((ScalarArgument) arguments.get("END")).getValue()).toStringUtf8();
+
+        if (Strings.isNullOrEmpty(query)) {
+            throw new TrinoException(INVALID_FUNCTION_ARGUMENT, query);
+        }
+        requireNonNull(startArgument);
+        requireNonNull(endArgument);
+        requireNonNull(metadata);
+
+        // determine the returned row type
+        List<ColumnHandle> columnHandles;
+        try {
+            columnHandles = metadata.getColumnHandles(query);
+        }
+        catch (UndeclaredThrowableException e) {
+            throw new TrinoException(INVALID_FUNCTION_ARGUMENT, "Cannot get column definition", Throwables.getRootCause(e));
+        }
+        if (columnHandles.isEmpty()) {
+            throw new TrinoException(INVALID_FUNCTION_ARGUMENT, "Cannot get column definition");
+        }
+        Descriptor returnedType = new Descriptor(columnHandles.stream()
+                .map(LokiColumnHandle.class::cast)
+                .map(column -> new Descriptor.Field(column.name(), Optional.of(column.type())))
+                .collect(toImmutableList()));
+
+        Instant end = toInstant(endArgument);
+        Instant start = toInstant(startArgument);
+        if (end.getEpochSecond() == 0) {
+            end = Instant.now();
+        }
+        if (start.getEpochSecond() == 0) {
+            start = end.minus(Duration.ofHours(1));
```

Comment:
That's not obvious for users. Maybe make `start` and `end` required arguments for the `query_range` function, and add a `query_last_hour` function without these?

Reply:
The Loki API has these as defaults. I think that's ok.

Reply:
Trino doesn't have to follow Loki's design decisions, and we try to be very explicit and avoid unexpected behavior. It's ok for this table function to be violatile and always return different results, but users should be aware of this. If the function name doesn't suggest that, at least document it somewhere. Neither functions nor their arguments can have a description, that's why I suggested a more descriptive name. This will definitely need to be called out in the connector docs.

Reply:
> we try to be very explicit and avoid unexpected behavior.

This makes a lot of sense. I'll make the parameters mandatory.